turbopack/crates/turbo-tasks-backend/src/backend/mod.rs

```diff
@@ -2162,20 +2165,50 @@ impl<B: BackingStorage> TurboTasksBackendInner<B> {
         cell_id: Option<CellId>,
     ) {
         let transient_reason = if let Some(child) = child {
-            format!(
+            Cow::Owned(format!(
                 " The callee is transient because it depends on:\n{}",
                 self.debug_trace_transient_task(child, cell_id),
-            )
+            ))
         } else {
-            String::new()
+            Cow::Borrowed("")
         };
         panic!(
-            "Persistent task {} is not allowed to call or read transient tasks {}.{}",
+            "Persistent task {} is not allowed to call, read, or connect to transient tasks {}.{}",
             parent.map_or("unknown", |t| t.get_name()),
             child.map_or("unknown", |t| t.get_name()),
             transient_reason,
         );
     }
+
+    fn assert_valid_collectible(&self, task_id: TaskId, collectible: RawVc) {
+        // these checks occur in a potentially hot codepath, but they're cheap
+        let RawVc::TaskCell(col_task_id, col_cell_id) = collectible else {
+            // This should never happen: The collectible APIs use
+            let task_info =
+                if let Some(col_task_ty) = self.lookup_task_type(collectible.get_task_id()) {
+                    Cow::Owned(format!(" (return type of {})", col_task_ty))
+                } else {
+                    Cow::Borrowed("")
+                };
+            panic!("Collectible{task_info} must be a ResolvedVc")
+        };
+        if col_task_id.is_transient() && !task_id.is_transient() {
+            let transient_reason =
+                if let Some(col_task_ty) = self.lookup_task_type(collectible.get_task_id()) {
+                    Cow::Owned(format!(
+                        ". The collectible is transient because it depends on:\n{}",
+                        self.debug_trace_transient_task(&col_task_ty, Some(col_cell_id)),
+                    ))
+                } else {
+                    Cow::Borrowed("")
+                };
+            // this should never happen
+            panic!(
+                "Collectible is transient, transient tasks cannot be emitted from persistent \
```

Comment:
Is this message really correct? Or should it rather say "transient collectibles cannot be emitted"?